collect_app/src/main/java/org/odk/collect/android/dao/helpers/ContentResolverHelper.java

```diff
@@ -76,20 +73,8 @@ public static String getFormPath(Uri uri) {
         return formPath;
     }
 
-    /**
-     * Using contentResolver to get a file's extension by the uri
-     *
-     * @param fileUri Whose name we want to get
-     * @return The file's extension without a dot eg. "mp3" not ".mp3"
-     */
-    public static String getFileExtensionFromUri(Context context, Uri fileUri) {
-        String fileName = new MediaUtils().getFileNameFromUri(context, fileUri);
-        if (fileName != null && fileName.contains(".")) {
-            return fileName.substring(fileName.lastIndexOf('.') + 1);
-        } else {
-            return fileUri.getScheme() != null && fileUri.getScheme().equals(ContentResolver.SCHEME_CONTENT)
-                    ? MimeTypeMap.getSingleton().getExtensionFromMimeType(context.getContentResolver().getType(fileUri))
-                    : MimeTypeMap.getFileExtensionFromUrl(fileUri.toString());
-        }
+    public static String getFileExtensionFromUri(Uri fileUri) {
+        String mimeType = getContentResolver().getType(fileUri);
+        return mimeType.substring(mimeType.lastIndexOf("/") + 1);
```

Comment:
Generally you are right and I tried it but in some cases (that seem pretty obvious) it returns null those cases are like:
- audio/mp3
- text/csv

Here https://stackoverflow.com/questions/12473851/how-i-can-get-the-mime-type-of-a-file-having-its-uri/12473985 where that solution is marked as accepted people also complain about the same.

Maybe I should use what you mentioned first and then in case of null my: `mimeType.substring(mimeType.lastIndexOf("/") + 1);`

or better yet (as a second step) I can read filename that should contain file extension:
```
        Cursor cursor = getContentResolver().query(fileUri, null, null, null, null);
        if (cursor != null && cursor.moveToFirst()) {
            name = cursor.getString(cursor.getColumnIndex(MediaStore.MediaColumns.DISPLAY_NAME));
        }
        String extension = name != null ? name.substring(name.lastIndexOf(".") + 1) : "";
```

this is what this library does https://github.com/anggrayudi/SimpleStorage. I know it's new and didn't get traction yet but i ran into it through this topic https://stackoverflow.com/questions/63080879/get-file-extension-using-android-storage-access-framework.

What do you think?

Reply:
ANDROID! Why must it always make simple things hard. Good catch on `MimeTypeMap#getExtensionFromMimeType` not actually working all the time.

Doing that first and then falling back seems good. I think you might actually want to have both fallbacks since neither is perfect -- `MediaStore.MediaColumns.DISPLAY_NAME` is not necessarily a file name. Maybe first use `MimeTypeMap#getExtensionFromMimeType`, then try the display column and then use the subtype?

Reply:
Ok I did it I even added another fourth aproach since I noticed that in case of selfie video only that last approach works well :scream:

Reply:
Yikes! Is that maybe because selfie video gives back a file URI (not a content URI) without a MIME type set? Maybe that's worth looking at again after this PR.